README
	   SGMLS.PM: A PERL5 CLASS LIBRARY FOR USE WITH THE
		       SGMLS AND NSGMLS PARSERS

			     Version 1.03


			   David Megginson,
			 University of Ottawa


[This is the fourth public release.  See "CHANGES", below, for a
summary of the most important changes since version 1.02.  For
detailed changes, see ChangeLog and DOC/ChangeLog; for known bugs, see
BUGS.]


This distribution contains SGMLS.pm, a perl5 class library for parsing
the output from James Clark's SGMLS and PSGMLS parsers.  This is a
free package, but it is _not_ Public-Domain software -- please see the
file COPYING in this directory for the terms of use.

Full documentation is available in SGML format the DOC directory, and
in Postscript, LaTeX, and HTML format in the various subdirectories of
the DOC directory.  Documentation in POD format is also included
within each of the modules, and once they are installed, you can view
them with perldoc (SGMLS, SGMLS::Ouptut, or SGMLS::Refs) -- there is
no POD documentation in sgmlspl yet.

I have included a Unix Makefile for installing the package on Unix and
Unix-like systems.  Make any necessary changes at the top of the
Makefile, then type

  make install

to install the perl5 scripts on your system.  If you want to test the
installation, change to the DOC subdirectory, read the comments at the
top of the make file there, then type

  make clean; make all

to regenerate the LaTeX and HTML documentation using the scripts which
you have just installed (you will have to have the SGMLS or NSGMLS
parser, the Docbook 2.2.1 DTD and SGML declaration, and the ISOpub and
ISOnum entity sets available on your system -- copies are included in
the DOC/Extras subdirectory).  You may also type

  make dvi

or

  make ps

to produce DVI or Postscript versions of the documentation for
printing.


** nb: Effective with v.1.02, SGMLS.pm supports the new output from
NSGMLS 0.4.13 with the -h option.


CHANGES

There have been some major changes since version 1.02.  File handles
passed as arguments to SGMLS::new should be qualified in the correct
package now.  The Output.pm module has been moved, so you should now
type

  use SGMLS::Output;

instead of

  use Output;

in all of your scripts.  There is also a new package, SGMLS::Refs, to
maintain a database of forward references from one processing pass to
another, like the LaTeX .aux files.  Once you have installed SGMLS.pm,
you can type

  perldoc SGMLS::Refs

at the shell prompt for more information, or you can read the new
section in the SGML documentation for sgmlspl.  All of the
documentation for this package is available online through my home
page, listed below.


Enjoy!

David Megginson


EMAIL: dmeggins@aix1.uottawa.ca
WWW: http://www.uottawa.ca/~dmeggins
Canada Post: Department of English,
             University of Ottawa,
             Ottawa, CANADA



ChangeLog
Sun Nov	 7 20:22:01 2010  Andy Bailey <GooseYArd@gmail.com>

* VERSION 1.1

	Updated the package to use Module::Build. This should help
	distribution packagers who commonly patched the old Makefile
	to rearrange the install.

	Applied Ardo van Rangelrooij's debian patches

Tue Dec  5 07:18:09 1995  David Megginson <david@baeda.english.uottawa.ca>

* VERSION 1.03ii

	* elisp/sgmls.el: Added my Gnu Emacs 19 mode for running (n)sgmls
	with a postprocessor.

	* Output.pm (push_output): For a 'handle' argument, force
	unqualified filehandles into the caller's package (reported by
	Rick Wong).

* VERSION 1.03

Sun Dec  3 11:37:08 1995  David Megginson  <david@baeda.english.uottawa.ca>

	* sgmlspl.pl (main::sgml): Stopped using SGMLS_Event::key method
	(why keep it around?).

	* SGMLS.pm: Added POD documentation to the module; changed the
	SGMLS_Event class constructor so that it does not take an explicit
	key argument (the key will be generated automatically if needed),
	and propagated the change throughout the module.

Sat Dec  2 21:24:05 1995  David Megginson  <david@baeda.english.uottawa.ca>

	* Refs.pm: Added POD documentation to the module.
	* Output.pm: Added POD documentation to the module.

Thu Nov 16 06:52:46 1995  David Megginson  <david@baeda.english.uottawa.ca>

	* Refs.pm: Refs::warn now returns 1 if a warning was printed, or 0
	if it was not.

Wed Nov 15 13:43:10 1995  David Megginson  <david@baeda.english.uottawa.ca>

	* Refs.pm: Added a new, simple package for managing references and
	reporting any changes.

	* SGMLS.pm: SGMLS::new now forces its filehandle argument into the
	caller's package, unless it is already qualified with a package
	name; reported by Rick Wong.

	* Makefile: Install modules in ${MODULEDIR} instead of ${PERL5DIR}
	-- that way, all add-ons can be qualified with SGMLS::.

	* Output.pm: Now uses Carp.pm and Exporter.pm to be a
	better-behavied module.  The symbols "output", "push_output", and
	"pop_output" will be exported into the caller's package rather
	than main::, and the caller will have some control over them.

	Changed the package from Output to SGMLS::Output.pm.

	* skel.pl: Use SGMLS::Output instead of Output.

	* sgmlspl.pl: Altered to create the SGMLS object with main::STDIN
	as the filehandle.  Have Output.pm export its symbols into main::
	rather than into SGMLS_PL::.

	Use SGMLS::Output instead of Output.

* VERSION 1.02

Thu Aug 24 10:00:52 1995  David Megginson  <david@baeda.english.uottawa.ca>

	* SGMLS.pm: Added SGMLS_Event::parse to retrieve the SGMLS parent
	class, SGMLS_Event::entity to look up an arbitrary entity, and
	SGMLS_Event::notation to look up an arbitrary notation.

	* Makefile (HTML_SOURCES): fixed faulty path, reported by Rick
	Wong (rwong@jessica.stanford.edu).

	* skel.pl: fixed bug in commented out 're' handler, reported by
	Rick Wong (rwong@jessica.stanford.edu).

	* SGMLS.pm: fixed bug in SGMLS_Attribute::set_value, reported by
        Rick Wong (rwong@jessica.stanford.edu).

	Loosened up checking of types in main loop, in case new types are
	added in the future.

	Added support for new output from NSGMLS with -h option: 'T' for
	external text entities, new entity types 'PI', 'TEXT', and 'ID'.

	Added support (not yet tested) for data entities.
	

* VERSION 1.01

Sat Aug 12 09:34:40 1995  David Megginson  <david@baeda.english.uottawa.ca>

	* Makefile: changed the order of the operations for make all, so
	that the necessary files are installed _before_ trying to make the
	documentation.  Also, introduced the target "dist" to make the
	most current version of all the documentation.

	* skel.pl: fixed the header comment for 'subdoc' handlers, and
	added default handlers for 'end_element', 'cdata', 're',
	'end_subdoc', and 'conforming' (reported by Rick Wong
	<rwong@jessica.stanford.edu>).

	* SGMLS.pm: fixed bugs in SGMLS_Attribute::set_value, cleaned up
	the recognition and processing of octal escape sequences in the
	(N)SGMLS output, and fixed a minor bug with generated-filenames in
	the external entities (reported by Rick Wong
	<rwong@jessica.stanford.edu>).

	* sgmls.pl: changed handler patterns to allow dashes and other
	characters in element, subdoc, and entity names, and fixed error
	in generic start_subdoc_handler (reported by Rick Wong
	<rwong@jessica.stanford.edu> and Alan Karben
	<karben@nrs.dowjones.com>).

	* Makefile: Corrected several typos in the "install_html" target
	(reported by Rick Wong <rwong@jessica.stanford.edu>).  Changed the
	perl binary location from /usr/bin/perl to /usr/local/bin/perl (I
	am working on a slightly different system now).

* VERSION 1.00 (initial release)


script/sgmlspl.pl
#!/usr/bin/perl
########################################################################
# sgmlspl: a simple SGML postprocesser for the SGMLS and NSGMLS
#          parsers (requires SGMLS.pm library).
#
# Copyright (c) 1995 by David Megginson <dmeggins@aix1.uottawa.ca>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
# $Log: sgmlspl.pl,v $
# Revision 1.8  1995/12/03  21:46:31  david
# Eliminated all use of the SGMLS_Event::key method.
#
# Revision 1.7  1995/11/15  20:22:24  david
# Changed "use Output" to "use SGMLS::Output".  Qualified the STDIN
# filehandle for the SGMLS object with the main:: package name.
#
# Revision 1.6  1995/08/12  16:15:14  david
# Revised version for 1.01 distribution.
#
# Revision 1.5  1995/04/27  11:52:25  david
# Changed 'print' to 'main::output' for re handler; empty string
# translates into an empty sub {} rather than a sub printing an empty
# string; instead of evaluating every argument as a perl script, take
# only the first as a perl script and the rest as its arguments; allow
# empty scripts or scripts which do not end with '1;'; pass the event
# itself as the second argument to each handler, after the event data.
#
# Revision 1.4  1995/04/23  14:44:58  david
# Use the Output package.  Fixed the $version variable.
#
# Revision 1.3  1995/04/22  21:02:49  david
# Added some missing 'last SWITCH;' statements in the sgmls function.
#
# Revision 1.2  1995/04/22  20:58:48  david
# Added $SGMLS_PL::version variable and changed SDATA notation from
# [SDATA] to |SDATA|.
#
# Revision 1.1  1995/04/22  14:40:50  david
# Initial revision
#
########################################################################

use SGMLS::Output;

package SGMLS_PL;
use SGMLS;

$version = '$Id: sgmlspl.pl,v 1.8 1995/12/03 21:46:31 david Exp $';

#
# Set up handler defaults.
#
$start_document_handler = sub {};
$end_document_handler = sub {};
$start_element_handlers = { '' => sub {} };
$end_element_handlers = { '' => sub {} };
$cdata_handler = sub { main::output($_[0]); };
$sdata_handlers = { '' => sub { main::output($_[0]);} };
$re_handler = sub { main::output("\n"); };
$pi_handler = sub { '' => sub {} };
$entity_handlers = { '' => sub {} };
$start_subdoc_handlers = { '' => sub {} };
$end_subdoc_handlers = { '' => sub {} };
$conforming_handler = sub {};

#
# Main access point: declare handlers for different SGML events.
#
# Usage: sgml(event, handler);
#
# The event may be one of the following strings, or a special pattern.
# The generic events are as follow:
#
#   'start'                 The beginning of the document.
#   'end'                   The end of the document.
#   'start_element'         The beginning of an element.
#   'end_element'           The end of an element.
#   'cdata'                 Regular character data.
#   'sdata'                 Special system-specific data.
#   're'                    A record-end.
#   'pi'                    A processing instruction.
#   'entity'                An external-entity reference.
#   'start_subdoc'          The beginning of a subdocument entity.
#   'end_subdoc'            The end of a subdocument entity.
#   'conforming'            The document is conforming.
#
# In addition to these generic events, it is possible to handlers
# for certain specific, named events, as follow:
#
#   '<GI>'                  The beginning of element GI.
#   '</GI>'                 The end of element GI.
#   '[SDATA]'               The system-specific data SDATA.
#   '&ENAME;'               A reference to the external entity ENAME.
#   '{ENAME}'               The beginning of the subdocument-entity ENAME.
#   '{/ENAME}'              The end of the subdocument-entity ENAME.
#
#
# The handler may be a string, which will simply be printed when the
# event occurs (this is usually useful only for the specific, named
# events), or a reference to an anonymous subroutine, which will
# receive two arguments: the event data and the event itself.  For 
# example,
#
#   sgml('<FOO>', "\n\\begin{foo}\n");
#
# and
#
#   sgml('<FOO>', sub { output("\n\\begin{foo}\n"); });
#
# will have identical results.
#
sub main::sgml {
    my ($spec,$handler) = (@_);
    if (ref($handler) ne 'CODE') {
	$handler =~ s/\\/\\\\/g;
	$handler =~ s/'/\\'/g;
	if ($handler eq '') {
	    $handler = sub {};
	} else {
	    $handler = eval "sub { main::output('$handler'); };";
	}
    }
  SWITCH: {
				# start-document handler
      $spec eq 'start' && do {
	  $start_document_handler = $handler;
	  last SWITCH;
      };
				# end-document handler
      $spec eq 'end' && do {
	  $end_document_handler = $handler;
	  last SWITCH;
      };
				# start-element handler
      $spec =~ /^<([^\/].*|)>$/ && do {
	  $start_element_handlers->{$1} = $handler;
	  last SWITCH;
      };
				# generic start-element handler
      $spec eq 'start_element' && do {
	  $start_element_handlers->{''} = $handler;
	  last SWITCH;
      };
				# end-element handler
      $spec =~ /^<\/(.*)>$/ && do {
	  $end_element_handlers->{$1} = $handler;
	  last SWITCH;
      };
				# generic end-element handler
      $spec =~ 'end_element' && do {
	  $end_element_handlers->{''} = $handler;
	  last SWITCH;
      };
				# cdata handler
      $spec eq 'cdata' && do {
	  $cdata_handler = $handler;
	  last SWITCH;
      };
				# sdata handler
      $spec =~ /^\|(.*)\|$/ && do {
	  $sdata_handlers->{$1} = $handler;
	  last SWITCH;
      };
				# generic sdata handler
      $spec eq 'sdata' && do {
	  $sdata_handlers->{''} = $handler;
	  last SWITCH;
      };
				# record-end handler
      $spec eq 're' && do {
	  $re_handler = $handler;
	  last SWITCH;
      };
				# processing-instruction handler
      $spec eq 'pi' && do {
	  $pi_handler = $handler;
	  last SWITCH;
      };
				# entity-reference handler
      $spec =~ /^\&(.*);$/ && do {
	  $entity_handlers->{$1} = $handler;
	  last SWITCH;
      };
				# generic entity-reference handler
      $spec eq 'entity' && do {
	  $entity_handlers->{''} = $handler;
	  last SWITCH;
      };
				# start-subdoc handler
      $spec =~ /^\{([^\/].*|)\}$/ && do {
	  $start_subdoc_handlers->{$1} = $handler;
	  last SWITCH;
      };
				# generic start-subdoc handler
      $spec eq 'start_subdoc' && do {
	  $start_subdoc_handlers->{''} = $handler;
	  last SWITCH;
      };
				# end-subdoc handler
      $spec =~ /^\{\/(.*)\}$/ && do {
	  $end_subdoc_handlers->{$1} = $handler;
	  last SWITCH;
      };
				# generic end-subdoc handler
      $spec eq 'end_subdoc' && do {
	  $end_subdoc_handlers->{''} = $handler;
	  last SWITCH;
      };
				# conforming handler
      $spec eq 'conforming' && do {
	  $conforming_handler = $handler;
	  last SWITCH;
      };

      die "Bad SGML handler pattern: $spec\n";
  }
}


#
# The first argument on the command line is a perl module which will be
# read here and evaluated in the 'main' package -- everything else will
# be an argument to it.
#
package main;

$ARGV = shift;
unless ($ARGV eq '' || do 'sgmlspl-specs/' . $ARGV || do $ARGV) {
    if (!-e $ARGV) {
	die "FATAL: $ARGV does not exist.\n";
    } elsif (!-r $ARGV) {
	die "FATAL: $ARGV exists but is read-protected.\n";
    } elsif ($@) {
	die "FATAL: $@\n";
    }
}


#
# Do the actual work, using the SGMLS package.
#
package SGMLS_PL;

$parse = new SGMLS(main::STDIN);	# a new parse object

&{$start_document_handler}();	# run the start handler.

				# run the appropriate handler for each
				# event
while ($event = $parse->next_event) {
    my $type = $event->type;
  SWITCH: {
      $type eq 'start_element' && do {
	  &{($start_element_handlers->{$event->data->name}||
		$start_element_handlers->{''} || sub {})}($event->data,$event);
	  last SWITCH;
      };
      $type eq 'end_element' && do {
	  &{($end_element_handlers->{$event->data->name}||
		$end_element_handlers->{''} || sub {})}($event->data,$event);
	  last SWITCH;
      };
      $type eq 'cdata' && do {
	  &{$cdata_handler}($event->data,$event);
	  last SWITCH;
      };
      $type eq 'sdata' && do {
	  &{($sdata_handlers->{$event->data}||
	     $sdata_handlers->{''} || sub {})}($event->data,$event);
	  last SWITCH;
      };
      $type eq 're' && do {
	  &{$re_handler}($event->data,$event);
	  last SWITCH;
      };
      $type eq 'pi' && do {
	  &{$pi_handler}($event->data,$event);
	  last SWITCH;
      };
      $type eq 'entity' && do {
	  &{($entity_handlers->{$event->data->name}||
	     $entity_handlers->{''} || sub {})}($event->data,$event);
	  last SWITCH;
      };
      $type eq 'start_subdoc' && do {
	  &{($start_subdoc_handlers->{$event->data->name}||
	     $start_subdoc_handlers->{''} || sub {})}($event->data,$event);
	  last SWITCH;
      };
      $type eq 'end_subdoc' && do {
	  &{($end_subdoc_handlers->{$event->data->name}||
	     $end_subdoc_handlers->{''} || sub {})}($event->data,$event);
	  last SWITCH;
      };
      $type eq 'conforming' && do {
	  &{$conforming_handler}($event->data,$event);
	  last SWITCH;
      };

      die "Unknown SGML event type: $type\n";
  }
}
				
&{$end_document_handler}();	# run the end handler
